### PR TITLE
Implement afterBlockExecute hook for DPoS Module - Closes #6722

### LINKED ISSUE #6722: Implement afterBlockExecute hook for DPoS module
### Description

Bundle 3 topics from their respective private functions in the hook:
- [x] #6874 
- [x] #6875 
- [x] #6876 

and implement necessary conditionals while implementing these functions(e.g. some portion of the code should work after bootstrap period or at the last block of round). Private functions must use reusable auxilary function that contains the actual logic of the topic.

Also update previousTimestampStore on successful forging and implement anything related to hook that is outside of reusable logic. And add unit tests. 

### Acceptance Criteria
- Unit tests must pass

### Additional Information
- Code must use new module APIs and as much as the old DPoS module code

### PR DESCRIPTION
**Pull request by mehmetegemen**
### What was the problem?

This PR resolves #6722 

### How was it solved?

- Added private DPoS module methods to afterBlockExecute adhering to required control flow.

### How was it tested?

- Unit tests created for the hook

### CODE REVIEW

File: framework/src/modules/dpos_v2/module.ts [hunk 1]
```diff
@@ -130,21 +130,26 @@ export class DPoSModule extends BaseModule {
...
 			EMPTY_KEY,
 			previousTimestampStoreSchema,
 		);
 		const { timestamp: previousTimestamp } = previousTimestampData;
+
 		await this._updateProductivity(context, previousTimestamp);
+
+		if (isLastBlockOfRound) {
+			await this._createVoteWeightSnapshot(context);
+			await this._updateValidators(context);
```
Comment (shuse2) [inline framework/src/modules/dpos_v2/module.ts:145]: update validators should only happen after init round. `nextRoundNumber > initRounds`

**Review by shuse2 — CHANGES_REQUESTED** [review]

File: framework/test/unit/modules/dpos_v2/module.spec.ts [hunk 2]
```diff
@@ -1176,4 +1177,136 @@ describe('DPoS module', () => {
...
+			it('should not create vote weight', () => {
+				expect(dpos['_createVoteWeightSnapshot']).toHaveBeenCalledTimes(0);
+			});
+
+			it('should not update validators', () => {
+				expect(dpos['_updateValidators']).toHaveBeenCalledTimes(0);
+			});
+		});
+
+		describe('when hook exits successfuly', () => {
```
Comment (ishantiw) [inline framework/test/unit/modules/dpos_v2/module.spec.ts:1279]:
```suggestion
		describe('when hook exits successfully', () => {
```

**Review by ishantiw — APPROVED** [review]

**Review by ishantiw — APPROVED** [review]

File: framework/test/unit/modules/dpos_v2/module.spec.ts [hunk 3]
```diff
@@ -1176,4 +1180,214 @@ describe('DPoS module', () => {
...
+				await previousTimestampStore.setWithSchema(
+					EMPTY_KEY,
+					{ timestamp: previousTimestamp },
+					previousTimestampStoreSchema,
+				);
+
+				await dpos.afterBlockExecute(context);
+			});
+
+			it('should create vote weight', () => {
```
Comment (shuse2) [inline framework/test/unit/modules/dpos_v2/module.spec.ts:1250]:
```suggestion
			it('should create vote weight snapshot', () => {
```

File: framework/test/unit/modules/dpos_v2/module.spec.ts [hunk 4]
```diff
@@ -1176,4 +1180,214 @@ describe('DPoS module', () => {
...
+				await previousTimestampStore.setWithSchema(
+					EMPTY_KEY,
+					{ timestamp: previousTimestamp },
+					previousTimestampStoreSchema,
+				);
+
+				await dpos.afterBlockExecute(context);
+			});
+
+			it('should not create vote weight', () => {
```
Comment (shuse2) [inline framework/test/unit/modules/dpos_v2/module.spec.ts:1284]:
```suggestion
			it('should not create vote weight snapshot', () => {
```

File: framework/test/unit/modules/dpos_v2/module.spec.ts [hunk 5]
```diff
@@ -1176,4 +1180,214 @@ describe('DPoS module', () => {
...
+				await previousTimestampStore.setWithSchema(
+					EMPTY_KEY,
+					{ timestamp: previousTimestamp },
+					previousTimestampStoreSchema,
+				);
+
+				await dpos.afterBlockExecute(context);
+			});
+
+			it('should create vote weight', () => {
```
Comment (shuse2) [inline framework/test/unit/modules/dpos_v2/module.spec.ts:1316]:
```suggestion
			it('should create vote weight snapshot', () => {
```

File: framework/test/unit/modules/dpos_v2/module.spec.ts [hunk 6]
```diff
@@ -1176,4 +1180,214 @@ describe('DPoS module', () => {
...
```
Comment (shuse2) [inline framework/test/unit/modules/dpos_v2/module.spec.ts:1350]:
```suggestion
			it('should create vote weight snapshot', () => {
```

**Review by shuse2 — APPROVED** [review]

File: framework/test/unit/modules/dpos_v2/module.spec.ts (hunk 3 above)
Comment (mehmetegemen) [reply framework/test/unit/modules/dpos_v2/module.spec.ts]: done

File: framework/test/unit/modules/dpos_v2/module.spec.ts (hunk 4 above)
Comment (mehmetegemen) [reply framework/test/unit/modules/dpos_v2/module.spec.ts]: done

File: framework/test/unit/modules/dpos_v2/module.spec.ts (hunk 5 above)
Comment (mehmetegemen) [reply framework/test/unit/modules/dpos_v2/module.spec.ts]: done

File: framework/test/unit/modules/dpos_v2/module.spec.ts (hunk 6 above)
Comment (mehmetegemen) [reply framework/test/unit/modules/dpos_v2/module.spec.ts]: done

File: framework/test/unit/modules/dpos_v2/module.spec.ts (hunk 2 above)
Comment (mehmetegemen) [reply framework/test/unit/modules/dpos_v2/module.spec.ts]: done

File: framework/src/modules/dpos_v2/module.ts (hunk 1 above)
Comment (mehmetegemen) [reply framework/src/modules/dpos_v2/module.ts:159]: done

**Pull request merged**